[bmalloc] Make SystemHeap enabled but unused by default

Achierius opened the pull request:
#### 0cdb20ed4dec7b1f38dbc0ef0e02530068746f07
<pre>
[bmalloc] Make SystemHeap enabled but unused by default
<a href="https://bugs.webkit.org/show_bug.cgi?id=302340">https://bugs.webkit.org/show_bug.cgi?id=302340</a>
<a href="https://rdar.apple.com/164497152">rdar://164497152</a>

Reviewed by Keith Miller.

Previously, libpas would delegate all allocations from the
bmalloc heap to the bmalloc SystemHeap whenever the SystemHeap was
enabled. In preparation for guard-objects, we will want to be able to
use the SystemHeap for a subset of allocations without turning it on for
the rest. This requires us to separate the notion of &apos;enablement&apos; from
that of &apos;using it to replace libpas for bmalloc allocations&apos;, which is
what this patch does.

Currently, the SystemHeap is always enabled, but is only used in the
same conditions that would have previously enabled it.
One new envvar has been added to configure this:
setting `WebKitSystemHeapOverrideEnablement=0` will disable the
SystemHeap entirely.

Canonical link: <a href="https://commits.webkit.org/303149@main">https://commits.webkit.org/303149@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/08cbabfd0e096d8bcc75898c031b90b2c26444c4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/131474 "3 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/3806 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/42439 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/138981 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/83249 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/edb4ff01-3b6c-4133-978d-3748f835948d) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/3830 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/3646 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/100393 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/83249 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/663d7223-6184-4048-a6a8-2d919e893c03) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/134420 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/2763 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/117684 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/81185 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/de65bae8-4cdd-4bf1-9e7d-32e408ef0414) 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/2677 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/419 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/82173 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/123499 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/111297 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/35810 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/141627 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/129928 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/3550 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/36325 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/108759 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/3598 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/3170 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/108986 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/27611 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/2695 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/114012 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/56738 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/3611 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/32423 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/162945 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/3436 "Built successfully") | | [⏳ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/JSC-ARMv7-32bits-Tests-EWS "Waiting to run tests") | | 
| | [❌ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/3637 "Hash 08cbabfd for PR 53756 does not build (failure)") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/3541 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

kmiller68 commented on Source/bmalloc/bmalloc/Allocator.cpp:
```diff
@@ -42,7 +42,7 @@ Allocator::Allocator(Heap& heap, Deallocator& deallocator)
     , m_deallocator(deallocator)
 {
-    BASSERT(!Environment::get()->isSystemHeapEnabled());
+    BASSERT(!Environment::get()->shouldBmallocAllocateThroughSystemHeap());
```
Weird, `Bmalloc` is not how I would have CamelCased `bmalloc` but it appears consistent.

kmiller68 commented on Source/bmalloc/bmalloc/Environment.cpp:
```diff
@@ -148,6 +148,12 @@ static bool isRunningOnValgrind()
+    return value ? !atoi(value) : false;
+}
+
 #if BUSE(CHECK_NANO_MALLOC)
```
An aside but I don't think Nano malloc exists anymore we can probably get rid of this.

Achierius replied:
Some complications, will look into it for another patch

kmiller68 reviewed (approved):
r=me

Achierius committed:
[bmalloc] Make SystemHeap enabled but unused by default
https://bugs.webkit.org/show_bug.cgi?id=302340
rdar://164497152

Reviewed by Keith Miller.

Previously, libpas would delegate all allocations from the
bmalloc heap to the bmalloc SystemHeap whenever the SystemHeap was
enabled. In preparation for guard-objects, we will want to be able to
use the SystemHeap for a subset of allocations without turning it on for
the rest. This requires us to separate the notion of 'enablement' from
that of 'using it to replace libpas for bmalloc allocations', which is
what this patch does.

Currently, the SystemHeap is always enabled, but is only used in the
same conditions that would have previously enabled it.
One new envvar has been added to configure this:
setting `WebKitSystemHeapOverrideEnablement=0` will disable the
SystemHeap entirely.

Canonical link: https://commits.webkit.org/303149@main
